---
title: "01_QC_preprocess"
author: "Victor Yuan"
date: "October 25, 2018"
output:
  html_document:
    keep_md: true
    toc: true
    toc_float: true
    toc_depth: 2
editor_options: 
  chunk_output_type: console
---

This script is for doing QC on the mset created from Rob Cox Marsit Fry and Michels data saved as mset_RCMFM_raw_eth.rds in C:/.../Robjects_final/

The QC that I do is mainly to decide on which samples to keep (any poor quality samples will be removed) and sites to filter against. I use a couple different strategies that I follow from Horvath (2013) age clock, Rachel's invariable probe paper (2016/2017), and the Roblab QC pipeline. I also do some sex analysis to infer sex on the Fry samples, and to also remove any samples with incorrect self-reported sex.

# 1.0 EDA

## 1.1 Load libraries and data

```{r Load libraries, message = FALSE, warning = FALSE}
library(minfi)
library(dplyr)
library(pheatmap)
library(ggplot2)
library(ggthemes)
library(reshape2)
library(gridExtra)
library(stringr)
library(waffle)
library(tidyr)
library(RColorBrewer) 
library(fastcluster)
library(parallelDist)
library(dendextend)
library(RColorBrewer)
library(caret)
library(pvclust)
library(parallel)
library(wateRmelon)
library(ape)
```

```{r load_data}
mset <- readRDS('../../Robjects_final/00_mset_RCMFM_raw_eth.rds')
pDat <- as.data.frame(pData(mset))
pDat$rownames <- rownames(pDat)
```

Here I check if my samples overlap with Sam's project:

```{r sam samples}
samdes <- read.csv('../../data/Robinson/Sam_des.txt')
head(samdes)

# remove Sam's samples without ethnicity info
samdes2 <- samdes %>% filter(Ethnicity != 'NA') %>% dplyr::select(ParticipantID, group, Ethnicity, GA, Sex)
dim(samdes2) # 35 samples

#how many of Sam's known ethnicity samples are in my combined dataset
sum(samdes2$ParticipantID %in% pDat$sampleNames) # 30/35

# which ones are they?
samdes3 <- samdes2[-which(samdes2$ParticipantID %in% pDat$sampleNames),] #what samples are not in pDat
samdes3

grep('PM139', pDat$sampleNames, value = T)
```

Above 5 samples present in Sam's samplesheet but not mine are because:
* PM139r2 and PM139r1 is in my dataset, just under a different label ('PM139_r2')
* PM158 no consent
* PM67 muslim
* PM66 sam can't remember but maintains this sample was removed with good reason

## 1.2 Dataset composition

```{r visualize_dataset}
# By dataset
table(pDat$Dataset)
ggplot(pDat, aes(pDat$Ethnicity, fill = pDat$Dataset)) +
  geom_histogram(stat = 'count') +
  theme_classic() +
  theme(axis.title.x = element_blank(),
        axis.ticks.x=element_blank()) +
  ylab('# samples') +
  labs(fill='Dataset\n') +
  #scale_fill_discrete(labels=c('Robinson' = "Robinson (n = 77)", 
  #                             'Cox' = "Cox (n = 46)", 
  #                             'Marsit' = "Marsit (n = 293)", 
  #                             'Fry' = "Fry (n = 27)", 
  #                             'Michels' = "Michels (n = 73)"))+
  scale_fill_tableau(labels=c('Robinson' = "Robinson (n = 77)", 
                               'Cox' = "Cox (n = 46)", 
                               'Marsit' = "GSE75248 (n = 289)", 
                               'Fry' = "GSE73375 (n = 24)", 
                               'Michels' = "GSE70453 (n = 72)")) +
  scale_y_continuous(breaks=seq(0,400,25)) +
  ggtitle('Combined data (n = 508)')

composition <- data.frame(Group = factor(c('Caucasian', 'Asian', 'African'))) %>%
                              mutate(SampleSize = c(396, 62, 60),
                                     Total = cumsum(SampleSize),
                                     midpoint = Total - SampleSize/2,
                                     label = paste(Group, '\n', round(SampleSize/sum(SampleSize) * 100, 2), '%'))
# make blank theme
blank_theme <- theme_minimal()+
  theme(
  axis.title.x = element_blank(),
  axis.title.y = element_blank(),
  panel.border = element_blank(),
  panel.grid=element_blank(),
  axis.ticks = element_blank(),
  plot.title=element_text(size=14, face="bold"),
  axis.text.x=element_blank(),
  axis.text.y = element_blank()
  )

# pie chart, by ethnicity              
ggplot(composition, aes(x = 1, weight = SampleSize, fill = Group)) +
  geom_bar(position = 'stack', width = 1) +
  coord_polar('y', start = 0) +
  geom_text(aes(x = 1.2, y = midpoint, label = label)) +
  blank_theme +
  theme(legend.title = element_blank())

#waffle chart, by ethnicity
SampleSize <- c(396, 62, 60)
label <- paste(composition$Group, ' (', round(SampleSize/sum(SampleSize) * 100, 2), '%)\n', sep = '')
names(SampleSize) <- label
waffle(SampleSize, rows = 20, flip = T, colors = c('darksalmon', 'gold1', 'olivedrab2')) +
  theme(legend.text = element_text(size = 24,vjust = -5),
        legend.key.size = unit(1.3, 'cm'))

raw <- getBeta(mset) #beta values
```

## 1.3 Sex Analysis

Here I predict sex and inspect sex chromosomesto identify sample mislabelling. I also use these tools to infer sex of the Fry samples.

```{r}
gset <- mapToGenome(ratioConvert(mset))

# plot predicted sex (Minfi)
ps <- getSex(gset, cutoff = -2)

ggplot(ps %>% as.data.frame(), aes(x = xMed, y = yMed)) +
  geom_point(aes(col = predictedSex))

# add predictedsex column
pDat$PredictedSex <- ps$predictedSex 

# add flagged column
pDat$Flag <- 'O'

# add inferred sex to 'sex' column for Fry/Michels (beecause we don't have their sex info)
pDat$Sex[which(pDat$Dataset == 'Fry')] <- pDat$PredictedSex[which(pDat$Dataset == 'Fry')]
pDat$Sex[which(pDat$Dataset == 'Michels')] <- pDat$PredictedSex[which(
  pDat$Dataset == 'Michels')]

pDat$Flag[which(pDat$Sex != pDat$PredictedSex)] <- 'X'

# Any samples with a predicted sex that doesn't match?
Flag <- pDat %>% filter(!(pDat$Sex == pDat$PredictedSex)) 
Flag

rm(gset)
```

3 samples appear to have incorrect sex information. I will further analyze these samples using sex chromosome analysis.

Note that amy found another sample with inconsistent sex information in the marsit data. I 
take a look at this sample.

Clustering on Sex Chromosomes

```{r cluster_onXY}
# Filter based on xy probes
ann450k = getAnnotation(IlluminaHumanMethylation450kanno.ilmn12.hg19)
sexCh <- ann450k$Name[ann450k$chr %in% c('chrX', 'chrY')] 

raw_XY <- raw[sexCh,]
raw_XY_Mar <- raw_XY[,pDat$Dataset == 'Marsit']
raw_XY_Cox <- raw_XY[,pDat$Dataset == 'Cox']
raw_XY_Rob <- raw_XY[,pDat$Dataset == 'Robinson']

# cluster sex probes
d <- dist(t(na.omit(raw_XY_Mar)))
hc_Mar <- hclust(d, method = "complete")
d <- dist(t(na.omit(raw_XY_Cox)))
hc_Cox <- hclust(d, method = "complete")
d <- dist(t(na.omit(raw_XY_Rob)))
hc_Rob <- hclust(d, method = "complete")

plot(as.phylo(hc_Mar), lab4ut="axial", type = "unrooted", edge.width=0.5, cex=0.8,
     tip.color=as.numeric(pDat$Sex)[pDat$Dataset=='Marsit']-1, main = 'Marsit')
plot(as.phylo(hc_Mar), lab4ut="axial", type = "unrooted", edge.width=0.5, cex=0.8,
     tip.color=as.numeric(as.factor(pDat$Flag))[pDat$Dataset=='Marsit'], main = 'Marsit')

plot(as.phylo(hc_Cox), lab4ut="axial", type = "unrooted", edge.width=0.5, cex=0.8,
     tip.color=as.numeric(pDat$Sex)[pDat$Dataset=='Cox']-1, main = 'Cox')
plot(as.phylo(hc_Cox), lab4ut="axial", type = "unrooted", edge.width=0.5, cex=0.8,
     tip.color=as.numeric(as.factor(pDat$Flag))[pDat$Dataset=='Cox'], main = 'Cox')

plot(as.phylo(hc_Rob), lab4ut="axial", type = "unrooted", edge.width=0.5, cex=0.8,
     tip.color=as.numeric(pDat$Sex)[pDat$Dataset=='Robinson']-1, main = 'Robinson')
plot(as.phylo(hc_Rob), lab4ut="axial", type = "unrooted", edge.width=0.5, cex=0.8,
     tip.color=as.numeric(as.factor(pDat$Flag))[pDat$Dataset=='Robinson'],
     main = 'Robinson')

# Find when a female/male labelled sample clusters with the opposite sex
sexMar <- (as.numeric(pDat %>% filter(Dataset == 'Marsit') %>% pull(Sex))*-1 + 4) 
psexMar <- cutree(hc_Mar, k = 2)
names(psexMar)[sexMar != psexMar]

ggplot( cbind(pDat, as.data.frame(ps)), aes(x = xMed, y = yMed)) +
  geom_point(aes(col = Sex)) +
  geom_text(aes(label = ifelse(sampleNames %in% names(psexMar)[sexMar != psexMar], sampleNames,'')))


# Generate heatmap, can choose to cluster rows / cols, and the labels (e.g. Sex, PredictedSex, Sample_Group)
cor_raw_XY <- cor(raw_XY, use = 'pairwise.complete.obs')

# Specify colors
ann_colors = list(
    Sex = c(M = "blue", F ="pink"),
    PredictedSex = c(M = "blue", F ="pink"),
    Dataset = brewer.pal(length(levels(pDat$Dataset)), 'Set1'),
    Ethnicity = c(Caucasian = '#984EA3', Asian = '#E41A1C', African = '#FFFF33', South_Asian = '#E41A1C', East_Asian = '#E41A1C'),
    Flag = c(O = 'Orange', X = 'Blue')
)

names(ann_colors$Dataset) <- levels(pDat$Dataset)

pheatmap(cor_raw_XY[,], cluster_rows=T, cluster_cols=T, 
         annotation_col = pDat %>% dplyr::select( Sex, PredictedSex, Dataset, Ethnicity, Flag), 
         annotation_colors = ann_colors,
         show_rownames = F, show_colnames = F,
         main = 'Clustering on Sex Chromosomes')
```

Samples with mislabelled sex: "GSM1947253" "GSM1947288" "GSM1947399"

# 2.0 Detection P BC

Load libraries and data 
```{r load data, message=FALSE, warning=FALSE}
lapply(paste('package:',names(sessionInfo()$otherPkgs),sep=""),detach,character.only=TRUE,unload=TRUE)
library(minfi)
library(wateRmelon)
library(dplyr)
library(pheatmap)
library(ggplot2)
library(RColorBrewer)
library(parallel)
library(gridExtra)
library(reshape)
library(RColorBrewer)

# load rob cox marsit data
rgset_RCM <- readRDS('../../Robjects_final/00_rgset_RobCoxMarsit.rds')
pDat_RCM <- as.data.frame(pData(rgset_RCM))

colnames(rgset_RCM) <- pDat_RCM$sampleNames
pDat_RCM$rownames <- rownames(pDat_RCM)
```

### 2.1 Load detp and merge

Here I load the detection p value data for each dataset and merge them all into one big data frame.
The Rob, Cox, and Marsit data has already been merged, when I combined them as one 'rgset'. 

```{r load detect p matrix}
# get detection p value matrix for RCM
system.time(det_RCM <- detectionP(rgset_RCM))

#remove extra ethn samples
ind <- pDat %>% filter(Dataset == 'Robinson' | Dataset == 'Cox' | Dataset == 'Marsit') %>%
  pull(sampleNames)
all(ind %in% colnames(det_RCM))
det_RCM <- det_RCM[,ind] 

# load fry detect p value
det_F <- readRDS('../../Robjects_final/00_detectP_Fry.rds')
pDat_Fry <- pDat %>% filter(Dataset == 'Fry')

# put samplenames into colnames of det_F
all(pDat_Fry$sampleNames %in% colnames(det_F))
det_F <- det_F[,pDat_Fry$sampleNames] #reorder det_F matching pDat_Fry
colnames(det_F) <-pDat_Fry$sampleNames
head(det_F)

# Load Michels
load('../../data/Michels/GSE70453 for Ethnicity.RData')
all(colnames(lumiMethy.GDM) %in% pDat$sampleNames) # FALSE, because of extra ethn samples

# remove samples with irrelevant ethnicity samples
ncol(lumiMethy.GDM) #82
lumiMethy.GDM <- lumiMethy.GDM[,pDat[pDat$Dataset=='Michels','sampleNames']]
ncol(lumiMethy.GDM) #73
all(colnames(lumiMethy.GDM) %in% pDat$sampleNames) # should be TRUE now

#get detection p values
det_Mic <- pvals(lumiMethy.GDM)

# get nbead info for A and B probes
bcA_Mic <- lumiMethy.GDM@assayData$Avg_NBEADS_A
bcB_Mic <- lumiMethy.GDM@assayData$Avg_NBEADS_B

# get indices which probes are <3
indA <- bcA_Mic < 3
indB <- bcB_Mic < 3

# combine into one
indA[indB] <- TRUE

sum(indA) # 33374 cells with bc < 3
bc_Mic <- indA   # so bc_Mic is a bead count matrix with logical values, where TRUE corresponds to a bead count atha is less than 3, and FALSE is for cells that have a bead count > 3
rm(indA, indB)

all(colnames(bc_Mic) %in% pDat$sampleNames)
```


Here I look at bad detection p value probes. I can get detection p values from the rgset containing Marsit, Cox, and Robinson datasets. 

```{r format data}
dim(det_F)   # 485577 27
dim(det_RCM) # 485512 418
dim(det_Mic) # 485577 73

det_F <- det_F[rownames(det_RCM),] # re order probes, drops the snp probes
det_Mic <- det_Mic[rownames(det_RCM),] # re order probes, drops the snp probes

identical(rownames(det_F), rownames(det_RCM)) #sanity check
identical(rownames(det_Mic), rownames(det_RCM)) #sanity check
identical(rownames(det_Mic), rownames(det_F)) #sanity check

dim(det_F)   # 485512 27
dim(det_RCM) # 485512 418
dim(det_Mic) # 485512 73

#make data frame
det_F <- as.data.frame(det_F)
det_RCM <- as.data.frame(det_RCM)
det_Mic <- as.data.frame(det_Mic)

# add merging column
det_F$Cpg <- rownames(det_F) # for merging
det_RCM$Cpg <- rownames(det_RCM)
det_Mic$Cpg <- rownames(det_Mic) # for merging

#merge
det <- inner_join(det_F, det_Mic, by = 'Cpg') # join fry and Michels
det <- inner_join(det, det_RCM, by = 'Cpg') # join FM and rcm

# drop Cpg column, input as rownames
rownames(det) <- det$Cpg
det <- det[, -match('Cpg', colnames(det))]
dim(det) #485512 518

# make sure pDat is in right order as det
identical(rownames(pDat), colnames(det)) #False
all(pDat$sampleNames %in% colnames(det)) # True - just in wrong order
ncol(det) #518
nrow(pDat) #518

# reorder
pDat <- pDat[colnames(det), ]
identical(rownames(pDat), colnames(det)) # re ordered
dim(pDat) #518 21
```

```{r save detection p, eval = FALSE}
saveRDS(det, "../../Robjects_final/01_detP_RCMFM.rds")
```

Detection p - sample filtering

Here I look at the number of failed probes defined as having a detection p value >0.01 in > 5% of all probes per sample, for the purpose of sample filtering.

```{r mean_detp}
barplot(colMeans(det),col=brewer.pal(8,"Dark2")[factor(pDat$Dataset)],
        las=2,cex.names=0.8, main="Mean detection p-values", ylim = c(0,0.02), 
        names.arg ='', xlab = 'Sample')
```


```{r}
# num of probes in a sample that fail the cutoff
ndet <- colSums(det > 0.01, na.rm = T)
head(ndet)
pDat$ndet <- ndet[match(pDat$rownames, names(ndet))] # put into pDat
```

Below I calculate the number of samples with a total number of bad detection p value probes that is greater than 5% of the total amount of probes (~4800). 

I also want to highlight samples that have a low mean interarray correlation (<0.95).

```{r greater_than_five_percent}
# how many samples have > 1% of their probes that failed the cutoff?
sum(pDat$ndet > 0.05*nrow(det)) # 0
```

Looks like no samples have >5% failed probes. 

ndetect p probes

Horvath only keeps probes with < 10 missing values (note)

```{r list of probes 5 missing values, eval = T}
probes_morethan5NAs <- as.data.frame(names(which((rowSums(det > 0.01, na.rm = T)>5))))
```

```{r save list of bad probes, eval = FALSE}
write.table(probes_morethan5NAs, '../../Robjects/probes_morethan5NAs.txt')
```

## 2.2 nbeads

Here I look at beadcount information. For this I only have information for the robinson, cox, 
marsit, and Michels data.

Examine thresholds

```{r determine_bead_count_threshold}
#obtain beadcounts from an extended rgset using Watermelon function
rgset_RCM <- rgset_RCM[,ind]
bc_RCM <- beadcount(rgset_RCM)
dim(bc_RCM) # 485512 418
sum(is.na(bc_RCM)) #237040

dim(bc_Mic) # 485577 73
bc_Mic <- bc_Mic[rownames(bc_RCM),]
dim(bc_Mic) # 485512 73
sum(bc_Mic) #33374

bc <- cbind(bc_RCM, bc_Mic)
dim(bc) #485512 491
head(bc) 

# change FALSE to '4', and TRUE|na to '2'
bc[bc == TRUE] <- 2
bc[bc == FALSE] <- 4
bc[is.na(bc)] <- 2
head(bc)

sum(is.na(bc)) # 0
sum(bc < 3) # how many signals have <3 beads contributing to signal? 270414

isna_bc <- bc < 3
numNA_probe <- rowSums(isna_bc) # how many samples with <3 bead counts for each probe
sum(numNA_probe > 5) # number of probes with <3 bead counts in > 5 samples, 5253

thresholds3 <- data.frame(cutoff = seq(1, 20, 1))
system.time(thresholds3 <- thresholds3 %>% mutate(
  Num_probes_remove = apply(thresholds3, 1, function(x) sum(numNA_probe > x))))

p7 <- ggplot(thresholds3, aes(x = cutoff, y = Num_probes_remove)) +
  geom_bar(stat='identity')+
  theme_classic() +
  xlab('Bead count < 3 in > x number of samples') +
  ylab('resulting # probes to remove') +
  ggtitle('Probe filtering on bead count') +
  scale_x_continuous(breaks = pretty(thresholds3$cutoff, n = 10)) +
  scale_y_continuous(breaks = pretty(thresholds3$Num_probes_remove, n = 8))+
  theme(panel.grid.major.y = element_line(size = 0.1, colour = 'gray')) +
  geom_vline(xintercept = 4, col = 'red')
p7
```

## 2.3 number of bad detect and bc probes

Here I combine bad detection p and bc information into one matrix. cells with EITHER bad detection p value >0.01 or bc < 3 are replaced with a '1', otherwise cells will be '0', indicating good quality.

Ultimately, I use this matrix to filter probes and determine low quality samples. Except for Fry, which only has detection p values available, and not bead count.

```{r}
all(colnames(bc) %in% colnames(det)) # TRUE
all(colnames(det) %in% colnames(bc)) # FALSE - b/c Fry doesn't contain beadcount info

fp <- det

dim(fp) # 485512  518
dim(bc) # 485512 491
dim(det)#  485512 518

# replace bad detection p and low bead count features with NA
fp[fp > 0.01] <- NA 
sum(is.na(fp)) # 300594

# subset out Fry info
fp_temp <- fp[,colnames(bc)]
dim(fp_temp) # 491 samples
fp_temp[bc < 3] <- NA
sum(is.na(fp_temp)) # 554091

# put back in
fp[,colnames(fp_temp)] <- fp_temp
dim(fp) #518 samples
sum(is.na(fp)) # 562933 562976

# replace failed values (NAs) with TRUE, and non failed values with FALSE
fp[!is.na(fp)] <- FALSE
fp[is.na(fp)] <- TRUE
sum(fp) # 562933 562976

fpt <- rowSums((fp)) # number of failed observations for each cpg

thresholds4 <- data.frame(cutoff = seq(1, 20, 1))
system.time(thresholds4 <- thresholds4 %>% mutate(
  Num_probes_remove = apply(thresholds4, 1, function(x) sum(fpt > x)))) # number of probes that have > x number of failed probes

p8 <- ggplot(thresholds4, aes(x = cutoff, y = Num_probes_remove)) +
  geom_bar(stat='identity')+
  theme_classic() +
  xlab('Bead count < 3 and bad detect p > 0.01 in > x number of samples') +
  ylab('resulting # probes to remove') +
  ggtitle('Probe filtering on failed probes') +
  scale_x_continuous(breaks = pretty(thresholds3$cutoff, n = 5)) +
  scale_y_continuous(breaks = pretty(thresholds3$Num_probes_remove, n = 4))+
  theme(panel.grid.major.y = element_line(size = 0.1, colour = 'gray')) +
  geom_vline(xintercept = 4, col = 'red')
p8
```

fpt - contains # of failed probes (both bad detect p and beadcount) in a p by 1 matrix
fp - contains NAs for failed observations, in a p by m matrix.

# 3.0 Pre-normalization filtering

Here I summarize the results for my QC analysis and filter samples and probes.

## 3.1 Setup

```{r libraries, message=FALSE, warning=FALSE, paged.print=FALSE}
lapply(paste('package:',names(sessionInfo()$otherPkgs),sep=""),detach,character.only=TRUE,unload=TRUE)

library(irlba) # fast PCA
library(rsvd) # fast (randomized) PCA
library(minfi)
library(dplyr)
library(pheatmap)
library(ggplot2)
library(RColorBrewer)
library(reshape)
library(gridExtra)
library(ggfortify)
library(RCurl)
library(parallel)
library(parallelDist)
library(dendextend)
library(waffle)
library(wateRmelon)

# check data
dim(mset) # 485577    518
dim(pDat) # 518  25
all(rownames(pDat) %in% colnames(mset)) #TRUE
identical(rownames(pDat), colnames(mset)) #FALSE
pDat <- pDat[colnames(mset),] # reorder to match

pData(mset) <- DataFrame(pDat)
raw_b <- getBeta(mset)
```

## 3.2 Probe Filtering

bad detection p and beadcount probes

Here I load a list of counts per CpG corresponding to the number of observations that failed detection p value or had a bead count less than 3.

I pull out from this list a new list of probes that had a count greater than 1% of the dataset (>5 samples).

```{r load probe lists}
# number of bad observarions per probe
badprobe <- fpt
head(badprobe)
badprobe<- names(badprobe)[badprobe > 0.01*ncol(mset)]
length(badprobe) # 14858
```

Cross hybridizing probes

Here I combine Magda's and Chen's annotations to form a list of 29770 cross hybridizing probes.

Magda's annotation (2013) has a list of CH probes. 

```{r Load Magda annotation}
library(GEOquery)
base <-'https://www.ncbi.nlm.nih.gov/geo/download/?acc=GPL16304&format=file'
end <- '&file=GPL16304%5FGene%5Ffeatures%5FPlatformTable%2Etxt%2Egz'

mprice <- as_tibble(getGEO('GPL16304')@dataTable@table)
mprice_XY <- mprice %>% filter(XY_Hits == 'XY_YES') %>% pull(ID)
mprice_AU <- mprice %>% filter(Autosomal_Hits == 'A_YES') %>% pull(ID)

mprice_CH <- union(mprice_XY, mprice_AU)
length(mprice_CH) #41937
```

Chen Y, Lemire M, Choufani S, Butcher DT, Grafodatskaya D, Zanke BW, Gallinger S, Hudson TJ, Weksberg R: Discovery of cross-reactive probes and polymorphic CpGs in the Illumina Infinium HumanMethylation450 microarray. Epigenetics 2013, 8:203-9.

```{r Load Chen annotation}
library(RCurl)
library(httr)
library(readxl)
base <- 'http://www.sickkids.ca/MS-Office-Files/Research/Weksberg%20Lab/'
end <- '48639-non-specific-probes-Illumina450k.xlsx'
GET((paste(base, end, sep = '')), write_disk(tf <- tempfile(fileext = ".xlsx")))
chen_CH <- read_excel(tf, 1L) %>% pull(TargetID)
length(chen_CH)
```

Now I combine the lists into one list of probes to filter

```{r combine filter probes}
CH <- union(mprice_CH, chen_CH)

length(badprobe) #14858
length(CH) #41993

filter.probes <- unique(c(badprobe, CH))
length(filter.probes) # 55637

badprobes <- which(rownames(mset) %in% filter.probes)
length(badprobes) # 55637
```

1) Probes that failed detection p value threshold (>0.01) and/or beadcount (<3) were marked as NAs.
Probes that had >1% of observations missing (0.01 * 518 samples ~=5) were removed.
This equated to 14858 probes that failed this QC check.

2) Cross hybridizing probes (autosome and XY) were removed. This was based on the combined annotations 
of Magda Price's 2013 publication, and from Chen's popular publication. 
This amounted to 41993 probes.

1) and 2) probesets overlap to a certain extent. But the overall total number of unique probes that
were filtered at this step is 55637.


```{r filter probes}
dim(mset) # 485577 518
mset <- mset[-badprobes,]
dim(mset) # 429940 518
```

## 3.3 Sample filtering

Interarray Sample-sample correlation

Next I look at mean sample-sample correlation. I mainly do this to identify outliers and to remove those that have too many technical artifacts or are of low quality. I draw upon Horvath from his age clock paper for inspiration. I also cite Rachel's nonvariable probe paper as they use this strategy as well.

```{r mean_sample_cor}
raw <- getBeta(mset)
cor_raw <- cor(raw, use = 'pairwise.complete.obs')

#Calculate mean sample-sample correlation
sample_cor <- apply(cor_raw, 1, mean)
sample_cor <- data.frame(mean_sample_cor = sample_cor, sampleName = as.factor(names(sample_cor)))
head(sample_cor)

#put in pData
pDat$meanSScor <- sample_cor[match(sample_cor$sampleName, rownames(pDat)),'mean_sample_cor']
```

Plotting by samples we can see if there are any low correlated samples. I put horizontal absolute value lines at 0.95, 0.955, and 0.96 to visualize the potential number of samples that will be removed at a respectively cutoff.

Horvath uses 0.90, but he was working with multiple tissue types. 

Rachel's data was much cleaner and within tissue. She could have used a cutoff of 0.95 (doesn't say in the paper, but there is a relevant figure).

```{r plot_samplsamplcor}
# This theme is for increasing text size
sizetheme <-  theme(axis.text.x = element_blank(),
        axis.text.y = element_text(size = 24),
        axis.ticks = element_blank(),
        text = element_text(size = 32),
        plot.title  = element_text(size = 36))

# Plot correlation by sample colored by chip
ggplot(pDat, aes(x = sampleNames, y = meanSScor, col = Dataset)) +
  geom_point() + 
  geom_text(aes(label = ifelse(meanSScor < 0.95, sampleNames, ''))) +
  geom_text(aes(label = ifelse(sampleNames == 'GSM1947201', sampleNames, '')), col = 'black') +
  scale_size(guide = FALSE) +
  scale_y_continuous(limits = c(0.88, 1.00)) +
  theme_classic() +
  theme(axis.text.x = element_blank(),
        axis.ticks.x = element_blank()) +
  scale_shape_manual(values = c(0, 8, 19)) +
  ggtitle('Mean sample-sample correlation') +
  labs(x = 'Sample', y = 'Correlation') +
  geom_hline(yintercept = 0.95, linetype = 'dashed', color = 'red') +
  geom_hline(yintercept = 0.955, linetype = 'dashed', color = 'orange') +
  geom_hline(yintercept = 0.96, linetype = 'dashed', color = 'brown')
```

Now I remove samples below 0.95:
  * DeID26     
  * Q4
  * Q16        
  * Q17        
  * GSM1947208

```{r print_bad_samples}
badsamps <- pDat %>% filter(meanSScor < 0.95) %>% 
  dplyr::select(sampleNames, rownames, meanSScor, Ethnicity, Dataset) %>% pull(sampleNames)
pData(mset) <- DataFrame(pDat) # put meanSScor into pData

mset <- mset[,setdiff(colnames(mset), badsamps)] # remove samples
dim(mset) #429940    513
```

Please see QCresults.xlsx for a table of all QC steps and questionable samples.

The following samples have mislabelled sex and will be removed:
  * GSM1947253
  * GSM1947288
  * GSM1947399

```{r}
sexmislabels <- c('GSM1947253', 'GSM1947288', 'GSM1947399')
mset <- mset[,setdiff(colnames(mset), sexmislabels)] # remove samples
dim(mset) #429940    510
pDat <- as.data.frame(pData(mset))
```

# 4.0 Normalization

Here I use BMIQ normalization. Previously, I had explored other methods.

```{r}
#serial
set.seed(1)
system.time(
  BMIQ <- as.data.frame(BMIQ(mset))
)
#  user   system  elapsed 
# 1659.796   13.608 1673.289 
```

# 5.0 Post-normalization filtering

Invariable probes

```{r warning=FALSE}
# Load the invariable probe data
library(RCurl) 
base <- paste("https://raw.githubusercontent.com/redgar598/",
        "Tissue_Invariable_450K_CpGs/master/", sep = "")
csv <- c("Invariant_Buccal_CpGs.csv", 
         "Invariant_Blood_CpGs.csv", 
         "Invariant_Placenta_CpGs.csv")
invar_Pl <- read.csv(text = getURL(paste(base, csv[3], sep = "")))$CpG

# remove invariable probes
Variation <- function(x) {quantile(x, c(0.9),na.rm=T)[[1]] -
        quantile(x, c(0.1), na.rm=T)[[1]]}
rr <- lapply(match(invar_Pl, rownames(BMIQ)),
                 function(x) Variation(BMIQ[x,]))
rr <- unlist(rr)

# get cpg names of probes with <0.05 range
iv <- intersect(invar_Pl[which(rr < 0.05)], invar_Pl)
length(iv) #86502

BMIQ <- BMIQ[setdiff(rownames(BMIQ), iv),]
dim(BMIQ) # 343438 510
```

Keep probes overlapping 450k/850k arrays

* previously I was usin epic B2 manifest
* now I use latest epic manifest, B4, rmeoving (319625 - 319223) more probes

```{r}
# load EPIC b4 manifest file
library(IlluminaHumanMethylationEPICanno.ilm10b4.hg19)
library(minfiData)
library(minfiDataEPIC)
annoEPIC <- getAnnotation(IlluminaHumanMethylationEPICanno.ilm10b4.hg19)
dim(annoEPIC) # 865859
dim(ann450k) # 485512 
cpg_ol <- intersect(annoEPIC$Name, ann450k$Name)
length(cpg_ol) # 452453 overlapping probes

data(RGsetEx)
data("RGsetEPIC")

# snp probes
snp_450k <- rownames(getSnpBeta(RGsetEx))
snp_EPIC <- rownames(getSnpBeta(RGsetEPIC))
length(snp_450k);length(snp_EPIC) #65 59
snp_ol <- intersect(snp_450k, snp_EPIC)
length(snp_ol) # 59

# remove non-overlapping probes from combined data
BMIQ_epic <- BMIQ[intersect(rownames(BMIQ), c(cpg_ol, snp_ol)),]
dim(BMIQ_epic) # 319233 510
```

# 6.0 Save processed data

```{r, eval = F}
saveRDS(BMIQ, '../../Robjects_final/01_processed_betas.rds')
saveRDS(BMIQ_epic, '../../Robjects_final/01_processed_betas_EPIC.rds')
saveRDS(pDat, '../../Robjects_final/01_pDat.rds')
```
